File: android/src/main/kotlin/jyahann/dgis_map_kit/MarkersController.kt
package jyahann.dgis_map_kit

import io.flutter.Log
import io.flutter.plugin.common.MethodChannel
import ru.dgis.sdk.map.MapObjectManager
import ru.dgis.sdk.map.Marker
import ru.dgis.sdk.map.SimpleMapObject
import java.util.Dictionary

class MarkersController(
    sdkContext: ru.dgis.sdk.Context,
    mapObjectManager: MapObjectManager,
    methodChannel: MethodChannel,
    layerId: String?
) {
    private var layerId: String?
    private var sdkContext: ru.dgis.sdk.Context
    private var mapObjectManager: MapObjectManager
    private var methodChannel: MethodChannel

    private var markersWithNoId: MutableList<Marker> = ArrayList();
    private var markersWithId: MutableMap<String, Marker> = mutableMapOf();


    init {
        this.layerId = layerId
        this.sdkContext = sdkContext
        this.methodChannel = methodChannel
        this.mapObjectManager = mapObjectManager
    }

    fun _getAllMarkers() : List<Marker> {
        return markersWithNoId.plus(markersWithNoId.map { marker -> marker });
    }

    fun addMarkers(markers: List<Any>) {
        val gisMarkers: MutableList<Marker> = ArrayList()

        for (marker in markers) {
            gisMarkers.add(_processNewMarker(marker))
        }

        mapObjectManager.addObjects(gisMarkers)
    }

    private fun _processNewMarker(marker: Any) : Marker {
        val markerId = (marker as Map<String, Any?>)["id"] as String?
        val marker = MarkersUtils.getMarkerFromDart(marker, this.sdkContext, layerId)
        if (markerId != null) {
            val oldMarker = getById(markerId)
            if (oldMarker != null) {
                mapObjectManager.removeObject(oldMarker)
            }
            markersWithId[markerId] = marker
        } else {
            markersWithNoId.add(marker)
        }
        return marker
    }

    fun addMarker(marker: Map<String, Any?>) {
        var marker = _processNewMarker(marker)

        mapObjectManager.addObject(marker)
    }


    fun getAll() : List<Any?> {
        return _getAllMarkers().map { marker -> (marker.userData as MapObjectUserData).userData }
    }

    fun getById(markerId: String) : Marker? {
        return markersWithId[markerId]
    }

    fun removeMarkerById(markerId: String) {
        val marker = getById(markerId)
        if (marker != null) {
            mapObjectManager.removeObject(marker)
            markersWithId.remove(markerId)
        } else {
            Log.d(
                "DGIS",
                "Marker on  remove: marker with given id ($markerId) isn't exists"
            )
        }
    }
    fun removeAll() {
        markersWithNoId = ArrayList()
        markersWithId = mutableMapOf()
        mapObjectManager.removeAll()
    }

    fun update(markerId: String, newMarker: Map<String, Any?>) {
        removeMarkerById(markerId)
        addMarker(newMarker)
    }

    fun close() {
        mapObjectManager.close()
    }
}

File: ios/Classes/MarkersController.swift
import DGis;
import Flutter;

class MarkersController {
    var layerId: String?;
    private var registrar: FlutterPluginRegistrar;
    private var sdk: DGis.Container;
    private var methodChannel: FlutterMethodChannel;
    private var map: DGis.Map;
    private let objectManager: MapObjectManager;
    
    private var markersWithId: Dictionary<String, Marker> = [:];
    private var markersWithNoId: Array<Marker> = [];
    
    init(layerId: String? = nil, registrar: FlutterPluginRegistrar, map: DGis.Map, sdk: DGis.Container, methodChannel: FlutterMethodChannel, objectManager: MapObjectManager) {
        self.layerId = layerId
        self.registrar = registrar
        self.map = map
        self.sdk = sdk
        self.methodChannel = methodChannel;
        self.objectManager = objectManager;
    }
    
    func getAllMarkers() -> Array<Marker> {
        return self.markersWithNoId + self.markersWithId.map{ $0.value };
    }
    
    func addMarkers(markers: Array<Any>) {
        var newMarkers = Array<Marker>();
        
        for marker in markers {
            newMarkers.append(
                _processNewMarker(marker: marker)
            );
        }
        
        NSLog("Adding markers on \(layerId ?? "")")
        
        self.objectManager.addObjects(objects: newMarkers);
    }
    
    private func _processNewMarker(marker: Any) -> Marker {
        let markerId = (marker as! Dictionary<String, Any?>)["id"] as? String;
        let marker = MarkersUtils.getMarkerFromDart(
            marker: marker as! Dictionary<String, Any?>,
            sdk: self.sdk,
            registrar: self.registrar,
            layerId: self.layerId
        );
        if (markerId != nil) {
            let oldMarker = getById(markerId: markerId!)
            if (oldMarker != nil) {
                self.objectManager.removeObject(item: oldMarker!)
            }
            markersWithId[markerId!] = marker
        } else {
            markersWithNoId.append(marker)
        }
        return marker
    }
    
    func addMarker(marker: Dictionary<String, Any?>) {
        self.objectManager.addObject(item: _processNewMarker(marker: marker));
    }
    
    func getAll() -> Array<Any?> {
        return self.getAllMarkers().map { ($0.userData as! MapObjectUserData).userData };
    }
    
    func getById(markerId: String) -> Marker? {
        return self.markersWithId[markerId];
    }
    
    func removeMarkerById(markerId: String) {
        let marker = getById(markerId: markerId)
        if marker != nil {
            self.objectManager.removeObject(item: marker!);
            self.markersWithId.removeValue(forKey: markerId);
        } else {
            NSLog("DGis: marker with given id \(markerId) isn't exists");
        }
    }
    
    func removeAll() {
        self.objectManager.removeAll();
    }
    
    func update(markerId: String, newMarker: Dictionary<String, Any?>) {
        removeMarkerById(markerId: markerId);
        addMarker(marker: newMarker);
    }
}
